app/models/hint.rb

```diff
@@ -22,7 +22,10 @@ class Hint < ApplicationRecord
 
   def self.match(searchterm)
     searchprint = fingerprint(searchterm)
-    Hint.find_by(fingerprint: searchprint)
+    hints = Hint.where(fingerprint: searchprint)
+    # Order matching hints by source; throw away nils; return first remaining
+    # Hint. If there are none, this will return nil.
+    sources.map { |source| hints.find_by(source: source) }.compact[0]
```

Comment:
I share your feeling about the query, and chose not to care because these querysets will be so small that the performance impact will be negligible, whereas the benefits in terms of semantics and configurability are large. (Although I still feel a bit guilty about the inefficiency.)